src/geometry.h

```diff
@@ -0,0 +1,1104 @@
+    Point3<T>& operator/=(T s) {
+        assert(s != 0);
+        return *this *= 1 / s;
+    }
+
+    std::string to_string() const {
+        return "(" + std::to_string(x) + ", " + std::to_string(y) + ", " +
+               std::to_string(z) + ")";
+    }
+
+public:
+    T x = 0;
+    T y = 0;
+    T z = 0;
+};
+
+using Point3f = Point3<float>;
+using Point3i = Point3<int>;
+
+template <typename T> Point3<T> operator*(T s, const Point3<T>& p) {
+    return {s * p.x, s * p.y, s * p.z};
+}
+
+template <typename T> float distance(const Point3<T>& p1, const Point3<T>& p2) {
+    return (p1 - p2).length();
+}
+
+template <typename T>
+float distance_squared(const Point3<T>& p1, const Point3<T>& p2) {
+    return (p1 - p2).length_squared();
+}
+
+/**
+ * Linearly interpolate between two points.
+ */
+template <typename T>
+Point3<T> lerp(float t, const Point3<T>& p0, const Point3<T>& p1) {
+    return (1 - t) * p0 + t * p1;
+}
+
+template <typename T> Point3<T> min(const Point3<T>& p1, const Point3<T>& p2) {
+    return {std::min(p1.x, p2.x), std::min(p1.y, p2.y), std::min(p1.z, p2.z)};
+}
+
+template <typename T> Point3<T> max(const Point3<T>& p1, const Point3<T>& p2) {
+    return {std::max(p1.x, p2.x), std::max(p1.y, p2.y), std::max(p1.z, p2.z)};
+}
+
+template <typename T> Point3<T> floor(const Point3<T>& p) {
+    return {std::floor(p.x), std::floor(p.y), std::floor(p.z)};
+}
+
+template <typename T> Point3<T> ceil(const Point3<T>& p) {
+    return {std::ceil(p.x), std::ceil(p.y), std::ceil(p.z)};
+}
+
+template <typename T> Point3<T> abs(const Point3<T>& p) {
+    return {std::abs(p.x), std::abs(p.y), std::abs(p.z)};
+}
+
+template <typename T>
+Point3<T> permute(const Point3<T>& p, int x, int y, int z) {
+    return {p[x], p[y], p[z]};
+}
+
+/**
+ * Point in 3-dimensional space.
+ */
+template <typename T> class Normal3 {
+public:
+    Normal3() = default;
+    Normal3(T x, T y, T z) : x{x}, y{y}, z{z} { assert(!contains_nan()); }
+
+    explicit Normal3(const Vector3<T>& v) : x(v.x), y(v.y), z(v.z) {
+        assert(!contains_nan());
+    }
+
+    template <typename U>
+    explicit Normal3(const Vector3<U>& v)
+        : x(static_cast<T>(v.x))
+        , y(static_cast<T>(v.y))
+        , z(static_cast<T>(v.z)) {
+        assert(!contains_nan());
+    }
+
+    template <typename U>
+    explicit Normal3(const Point3<U>& p)
+        : x(static_cast<T>(p.x))
+        , y(static_cast<T>(p.y))
+        , z(static_cast<T>(p.z)) {
+        assert(!contains_nan());
+    }
+
+    template <typename U> explicit operator Vector3<U>() const {
+        return {x, y, z};
+    }
+
+    bool contains_nan() const {
+        return std::isnan(x) || std::isnan(y) || std::isnan(z);
+    }
+
+    T operator[](size_t i) const {
+        assert(i < 3);
+        return i == 0 ? x : i == 1 ? y : z;
+    }
+
+    T& operator[](size_t i) {
+        assert(i < 3);
+        return i == 0 ? x : i == 1 ? y : z;
+    }
+
+    bool operator==(const Normal3<T>& n) const {
```

Comment:
Then, this will be a strict comparison, which is expected I guess. If we want to compare in an approximative way, we should have a function which clearly states it in its name.